frontend/src/hooks/useUndoRedo.tsx

```diff
@@ -0,0 +1,72 @@
+import { useRef, useCallback, useEffect } from "react";
+import { fabric } from "fabric";
+
+const useUndoRedo = (initialCanvas: fabric.Canvas | null) => {
+  const canvasRef = useRef<fabric.Canvas | null>(initialCanvas);
+  const undoStack = useRef<Array<any>>([]);
+  const redoStack = useRef<Array<any>>([]);
+
+  useEffect(() => {
+    canvasRef.current = initialCanvas;
+  }, [initialCanvas]);
+
+  const areStatesEqual = (state1: any, state2: any) => {
+    return JSON.stringify(state1) === JSON.stringify(state2);
+  };
+
+  const saveState = useCallback(() => {
+    const canvas = canvasRef.current;
+    if (canvas) {
```

Comment:
We could do an early return here. Something like
```suggestion
    if (!canvas) { return; }
    ...
```